chainstate/src/detail/chainstateref/mod.rs

```diff
@@ -391,55 +392,56 @@ impl<'a, S: BlockchainStorageRead, V: TransactionVerificationStrategy> Chainstat
+                        Ok(())
+                    }
+                }
+                OutputTimeLock::ForSeconds(_) => Err(
+                    CheckBlockError::InvalidBlockRewardMaturityTimelockType(block.get_id()),
+                ),
+            }
+        };
+
         for output in block.block_reward().outputs() {
             match output.purpose() {
-                common::chain::OutputPurpose::Transfer(_) => {
+                OutputPurpose::Transfer(_) => {
                     return Err(CheckBlockError::InvalidBlockRewardOutputType(
                         block.get_id(),
                     ))
                 }
-                common::chain::OutputPurpose::LockThenTransfer(_, tl) => match tl {
-                    common::chain::timelock::OutputTimeLock::UntilHeight(_) => {
-                        return Err(CheckBlockError::InvalidBlockRewardMaturityTimelockType(
-                            block.get_id(),
-                        ))
-                    }
-                    common::chain::timelock::OutputTimeLock::UntilTime(_) => {
-                        return Err(CheckBlockError::InvalidBlockRewardMaturityTimelockType(
-                            block.get_id(),
-                        ))
-                    }
-                    common::chain::timelock::OutputTimeLock::ForBlockCount(c) => {
-                        let cs: i64 = (*c)
-                            .try_into()
-                            .map_err(|_| {
-                                CheckBlockError::InvalidBlockRewardMaturityDistanceValue(
-                                    block.get_id(),
-                                    *c,
-                                )
-                            })
-                            .log_err()?;
-                        let given = BlockDistance::new(cs);
-                        if given < required {
-                            return Err(CheckBlockError::InvalidBlockRewardMaturityDistance(
-                                block.get_id(),
-                                given,
-                                required,
-                            ));
-                        }
-                    }
-                    common::chain::timelock::OutputTimeLock::ForSeconds(_) => {
-                        return Err(CheckBlockError::InvalidBlockRewardMaturityTimelockType(
-                            block.get_id(),
-                        ))
-                    }
-                },
-                common::chain::OutputPurpose::StakePool(_) => {
+                OutputPurpose::LockThenTransfer(_, tl) => timelock_check_fn(tl)?,
+                OutputPurpose::StakePool(_) => {
                     return Err(CheckBlockError::InvalidBlockRewardOutputType(
                         block.get_id(),
                     ))
                 }
-                common::chain::OutputPurpose::Burn => {
+                OutputPurpose::SpendStakePool(_, tl) => timelock_check_fn(tl)?,
```

Comment:
When staking, the limitation to spending shouldn't be applied to staking, but to decomissioning.

Meaning: If you stay, you can still stake with no limitation. But when you decommission, then you must way for a period of time.

Reply:
right